packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,453 @@
+           * Calls the guest function, either for the initial run or at the
+           * start of a replay.
+           *
+           * @param {boolean} [eager]
+           */
+          restart(eager = startEager) {
+            const { state, facets } = this;
+            const { activationArgs, log, bijection, outcomeKit } = state;
+            const { flow, admin, wakeWatcher } = facets;
+
+            const startFlowState = flow.getFlowState();
+
+            startFlowState !== 'Done' ||
+              // separate line so I can set a breakpoint
+              Fail`Cannot restart a done flow ${flow}`;
+
+            admin.reset();
+            if (eager) {
+              eagerWakers.add(flow);
+            }
+
+            const wakeWatch = vowish => {
+              // Extra paranoid because we're getting
+              // "promise watcher must be a virtual object"
+              // in the general vicinity.
+              zone.isStorable(vowish) ||
+                Fail`vowish must be storable in this zone (usually, must be durable): ${vowish}`;
+              zone.isStorable(wakeWatcher) ||
+                Fail`wakeWatcher must be storable in this zone (usually, must be durable): ${wakeWatcher}`;
+              watch(vowish, wakeWatcher);
+            };
+            const panic = err => admin.panic(err);
+            const membrane = makeReplayMembrane(
+              log,
+              bijection,
+              vowTools,
+              wakeWatch,
+              panic,
+            );
+            const eph = tmp.for(flow);
+            eph.membrane = membrane;
+            const guestArgs = membrane.hostToGuest(activationArgs);
+
+            const flowState = flow.getFlowState();
+            flowState === 'Running' ||
+              flowState === 'Replaying' ||
+              Fail`Restarted flow must be Running or Replaying ${flow}`;
+
+            // In case some host vows were settled before the guest makes
+            // the first call to a host object.
+            membrane.wake();
+
+            // We do *not* call the guestAsyncFunc by having the membrane make
+            // a host wrapper for the function. Rather, we special case this
+            // host-to-guest call by "manually" sending the arguments through
+            // and calling the guest function ourselves. Likewise, we
+            // special case the handling of the guestResultP, rather than
+            // as the membrane to make a host vow for a guest promise.
+            // To support this special casing, we store additional replay
+            // data in this internal flow instance -- the host activationArgs
+            // and the host outcome vow kit.
+            const guestResultP = (async () =>
+              // async IFFE ensures guestResultP is a fresh promise
+              guestAsyncFunc(...guestArgs))();
+
+            if (flow.getFlowState() !== 'Failed') {
+              bijection.init(guestResultP, outcomeKit.vow);
+            }
```

Comment:
It took me a minute to understand why the state check necessary. Let's add a comment that the bijection would have been reset already in case of a failure.

Reply:
Done